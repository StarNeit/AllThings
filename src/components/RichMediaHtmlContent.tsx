import React from 'react'
import SanitizedHtml from './SanitizedHtml'
import { css } from 'glamor'

/**
 * This component is used to display rich html content to users in a standardized
 * way. It also allows you to convert some rich media links like youtube into
 * embeds.
 */

interface IProps {
  allowYoutube?: boolean
  html: string
  localAnchors?: boolean
}

class RichMediaHtmlContent extends React.Component<IProps> {
  static defaultProps = {
    allowYoutube: false,
  }

  injectYoutubeVideos = (content: string) => {
    const youtubeRegex = /https:\/\/(youtu\.be\/|www.youtube\.com\/watch\?v=)([\w-]+)/gi
    const youtubeWrapper = `
      <div class="youtubeWrapper">
        <iframe src="https://youtube.com/embed/$2" frameborder="0" allowfullscreen></iframe>
      </div>
    `
    return content.replace(youtubeRegex, youtubeWrapper)
  }

  render() {
    const { html, allowYoutube, ...props } = this.props
    return (
      <SanitizedHtml
        localAnchors
        html={allowYoutube ? this.injectYoutubeVideos(html) : html}
        {...props}
        config={{
          ADD_ATTR: ['target'],
          ADD_TAGS: ['iframe'],
          FORBID_TAGS: ['style'],
          FORBID_ATTR: ['style'],
        }}
        {...css({
          '> ul': {
            paddingLeft: '40px',
            marginTop: '10px',
            marginBottom: '10px',
            listStyleType: 'disc',
          },
          ' a': { textDecoration: 'underline' },
          '> p': {
            margin: 0,
            marginBottom: '10px',
          },
        })}
      />
    )
  }
}

export default RichMediaHtmlContent
