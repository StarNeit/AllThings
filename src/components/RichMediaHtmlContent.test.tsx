import React from 'react'
import RichMediaHtmlContent from './RichMediaHtmlContent'
import { mount } from 'enzyme'
import { Provider } from 'react-redux'

const store = global.mockStore({
  authentication: {
    accessToken: 'abcd1234',
  },
  app: {
    embeddedLayout: false,
  },
})

describe('Check RichMediaHtmlContent', () => {
  it('should turn youtube links into embeds', () => {
    const content = mount(
      <Provider store={store}>
        <RichMediaHtmlContent
          allowYoutube
          html="ok https://www.youtube.com/watch?v=TESTID"
        />
      </Provider>,
    )
    expect(content.html()).toEqual(
      expect.stringContaining(
        '<iframe src="https://youtube.com/embed/TESTID"></iframe>',
      ),
    )
  })

  it('should not turn youtube links into embeds', () => {
    const html = `ok https://www.youtube.com/watch?v=TESTID`
    const content = mount(
      <Provider store={store}>
        <RichMediaHtmlContent html={html} />
      </Provider>,
    )
    expect(content.html()).toEqual(expect.stringContaining(html))
  })

  it('should apply css for injected html', () => {
    const content = mount(
      <Provider store={store}>
        <RichMediaHtmlContent
          allowYoutube
          html={`<ul><li>hi</li></ul><p>Test</p>`}
        />
      </Provider>,
    )
    expect(content).toMatchSnapshot()
  })

  it('should strip styles', () => {
    const content = mount(
      <Provider store={store}>
        <RichMediaHtmlContent
          allowYoutube
          html={`<style>html, body { background-color: "red"; } </style>hi</li></ul><p style="font-color: red">Test</p>`}
        />
      </Provider>,
    )
    expect(content.html()).toBe(
      `<div data-css-4p3ux5="" data-css-b00tz="">hi<p>Test</p></div>`,
    )
  })
})
