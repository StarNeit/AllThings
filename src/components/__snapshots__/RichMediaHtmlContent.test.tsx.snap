// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check RichMediaHtmlContent should apply css for injected html 1`] = `
.css-0> ul,
[data-css-0]> ul {
  padding-left: 40px;
  margin-top: 10px;
  margin-bottom: 10px;
  list-style-type: disc;
}

.css-0 a,
[data-css-0] a {
  text-decoration: underline;
}

.css-0> p,
[data-css-0]> p {
  margin: 0;
  margin-bottom: 10px;
}

.css-1,
[data-css-1] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

<Connect(SanitizedHtml)
  config={
    Object {
      "ADD_ATTR": Array [
        "target",
      ],
      "ADD_TAGS": Array [
        "iframe",
      ],
      "FORBID_ATTR": Array [
        "style",
      ],
      "FORBID_TAGS": Array [
        "style",
      ],
    }
  }
  data-css-0=""
  html="<ul><li>hi</li></ul><p>Test</p>"
  localAnchors={true}
>
  <ForwardRef
    dangerouslySetInnerHTML={
      Object {
        "__html": "<ul><li>hi</li></ul><p>Test</p>",
      }
    }
    data-css-0=""
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<ul><li>hi</li></ul><p>Test</p>",
        }
      }
      data-css-1=""
      data-css-0=""
    />
  </ForwardRef>
</Connect(SanitizedHtml)>
`;
